### DISCUSSION TITLE
Choosing CPU sampling interval

### DISCUSSION THREAD

**Discussion by sandbubbles** [plasma-umass/scalene#801] [Q&A] [answered]
Hi everyone,
Scalene uses a default sampling interval of 10 milliseconds. 
Could someone from the Scalene development team shed some light on why 10 milliseconds was chosen as the default sampling interval? 
Are there any recommendations or considerations for adjusting the sampling interval? For example, is there some benchmark that would indicate what is a suitable sampling interval?

**Comment by emeryberger** [comment, marked as answer]
10ms is pretty standard for sampling profilers: it's the default interval used by [Go](https://sumercip.com/posts/inside-the-go-cpu-profiler/), [VTune](https://www.intel.com/content/www/us/en/docs/vtune-profiler/user-guide/2024-0/minimizing-collection-overhead.html), and the [async profiler](https://github.com/async-profiler/async-profiler) for Java. [Py-spy](https://github.com/benfred/py-spy) also uses a 10ms rate. We also used 10ms in our [Coz](https://github.com/plasma-umass/coz) causal profiler.

This rate empirically strikes a sweet spot providing high accuracy and low overhead (for Scalene, see the graph of Figure 5 [from our OSDI paper](https://www.usenix.org/system/files/osdi23-berger.pdf) demonstrating that Scalene's choice delivers high accuracy).

There are a few notable departures from this default: [v8](https://chromium.googlesource.com/v8/v8/+/refs/heads/main/include/v8-profiler.h)'s profiler uses a 1ms interval. I just tried [perf](https://perf.wiki.kernel.org/index.php/Tutorial#Period_and_rate) and while the linked page says it uses a 1ms interval, on my system, it was more like 0.2ms. [Austin](https://github.com/P403n1x87/austin) also uses a default of 0.1ms.

**Reply by sandbubbles** [reply]
Thank you for the explanation!
Am I correct in understanding that this decision doesn't necessarily need to consider factors such as the speed at which GIL interprets instructions or Scalene's maximum processing pace for signals and samples?

Furthermore, Scalene adds to both counters at each sample - for Python execution, it is increased by the sampling interval and for native execution, it is increased by the difference between the elapsed time and the sampling interval. Is this approach designed to compensate for the lack of precise knowledge about the exact moment of transition from Python execution to native code?

**Reply by emeryberger** [reply]
First, a clarification: the GIL is just a lock (the _G_lobal _I_nterpreter _L_ock); it's the Python interpreter that interprets bytecodes. 

Second, Scalene actually only samples at a rate of _at most_ once per 10ms. It restarts the timer after it processes each sample. This implicitly adjusts for processing time.

And yes, its use of two separate counters lets it tease apart time spent in the Python interpreter vs. time spent in native code.

**Reply by emeryberger** [reply]
I should add that the latter part is covered in detail in [this technical paper (OSDI 2023)](https://www.usenix.org/system/files/osdi23-berger.pdf).
